tests/unittests/Foundation/NSOperationTests.mm

```diff
@@ -59,3 +62,256 @@
+
+    [operation addObserver:observer forKeyPath:@"completionBlock" options:0 context:NULL];
+    [operation addObserver:observer forKeyPath:@"isCancelled" options:0 context:NULL];
+    [operation addObserver:observer forKeyPath:@"isExecuting" options:0 context:NULL];
+    [operation addObserver:observer forKeyPath:@"isFinished" options:0 context:NULL];
+
+    [operation setCompletionBlock:^{
+        // nothing to do here.
+    }];
+
+    ASSERT_TRUE([observer didObserveCompletionBlock]);
+    [observer setDidObserveCompletionBlock:NO];
+
+    ASSERT_FALSE([observer didObserveCompletionBlock]);
+    ASSERT_FALSE([observer didObserveCancelled]);
+    ASSERT_FALSE([observer didObserveExecuting]);
+    ASSERT_FALSE([observer didObserveFinished]);
+
+    [queue addOperation:operation];
+    [operation waitUntilFinished];
+
+    ASSERT_TRUE([observer didObserveCompletionBlock]);
+    ASSERT_FALSE([observer didObserveCancelled]);
+    ASSERT_TRUE([observer didObserveExecuting]);
+    ASSERT_TRUE([observer didObserveFinished]);
+
+   [operation removeObserver:observer forKeyPath:@"completionBlock" context:NULL];
+   [operation removeObserver:observer forKeyPath:@"isCancelled" context:NULL];
+   [operation removeObserver:observer forKeyPath:@"isExecuting" context:NULL];
+   [operation removeObserver:observer forKeyPath:@"isFinished" context:NULL];
+}
+
+// Test asynchronous subclass for NSOperation
+@interface MyConcurrentOperation : NSOperation
+
+@property (assign, getter = isExecuting) BOOL executing;
+@property (assign, getter = isFinished, readonly) BOOL finished;
+
+@end
+
+@implementation MyConcurrentOperation
+
+@synthesize executing = _executing;
+@synthesize finished = _finished;
+
+- (void)setExecuting:(BOOL)executing
+{
+    [self willChangeValueForKey:@"isExecuting"];
+    [self willChangeValueForKey:@"isFinished"];
+
+    _executing = executing;
+    _finished = !executing;
+
+    [self didChangeValueForKey:@"isExecuting"];
+    [self didChangeValueForKey:@"isFinished"];
+}
+
+- (BOOL)isExecuting
+{
+    return _executing;
+}
+
+- (BOOL)isFinished
+{
+    return _finished;
+}
+
+- (void)start
+{
+    if (self.isCancelled) {
+        return;
+    }
+
+    self.executing = YES;
+    [self doSomething];
+}
+
+- (void)doSomething
+{
+    // Do some async task.
+    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_LOW, 0), ^{
+
+        // Do another async task
+        dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_LOW, 0), ^{
+            // Do some async task.
+            self.executing = NO;
+        });
+    });
+}
+
+@end
+
+TEST(NSOperation, NSOperationConcurrentSubclass) {
+    NSOperationQueue* queue = [[NSOperationQueue new] autorelease];
+
+    NSOperation* operation = [MyConcurrentOperation new];
+
+    [operation setCompletionBlock:^{
+        [operation waitUntilFinished]; // Should not deadlock, but we cannot test this
+        ASSERT_TRUE([operation isFinished]);
```

Comment:
We can somewhat test deadlocks -- by having a watchdog thread. In the worst case, the test runs until the watchdog fails it, but then we need to abort the test run completely since we can't recover.